File: prisma/refs/pacific-atlantic-water-flow.ts
export function pacificAtlantic(heights: number[][]): number[][] {
  const m = heights.length, n = heights[0]?.length ?? 0;
  if (!m || !n) return [];
  const mk = () => Array.from({ length: m }, () => new Array<boolean>(n).fill(false));
  const pac = mk(), atl = mk();
  const dfs = (r: number, c: number, seen: boolean[][]) => {
    seen[r][c] = true;
    for (const [dr, dc] of [[1, 0], [-1, 0], [0, 1], [0, -1]]) {
      const nr = r + dr, nc = c + dc;
      if (nr < 0 || nr >= m || nc < 0 || nc >= n) continue;
      if (seen[nr][nc] || heights[nr][nc] < heights[r][c]) continue;
      dfs(nr, nc, seen);
    }
  };
  for (let r = 0; r < m; r++) {
    dfs(r, 0, pac);
    dfs(r, n - 1, atl);
  }
  for (let c = 0; c < n; c++) {
    dfs(0, c, pac);
    dfs(m - 1, c, atl);
  }
  const out: number[][] = [];
  for (let r = 0; r < m; r++)
    for (let c = 0; c < n; c++) if (pac[r][c] && atl[r][c]) out.push([r, c]);
  return out;
}

File: prisma/refs/py/pacific-atlantic-water-flow.py
def pacific_atlantic(heights):
    m, n = len(heights), len(heights[0]) if heights else 0
    if not m or not n:
        return []

    def dfs(r, c, seen):
        seen.add((r, c))
        for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
            nr, nc = r + dr, c + dc
            if 0 <= nr < m and 0 <= nc < n and (nr, nc) not in seen \
                    and heights[nr][nc] >= heights[r][c]:
                dfs(nr, nc, seen)

    pac, atl = set(), set()
    for r in range(m):
        dfs(r, 0, pac)
        dfs(r, n - 1, atl)
    for c in range(n):
        dfs(0, c, pac)
        dfs(m - 1, c, atl)
    return [[r, c] for r in range(m) for c in range(n) if (r, c) in pac and (r, c) in atl]
